nucliadb/src/nucliadb/ingest/fields/base.py

```diff
@@ -596,3 +596,101 @@ async def set_value(self, value: Any):
 
     async def get_value(self) -> Any:
         raise NotImplementedError()
+
+
+class ExtractedTextIndex:
+    def __init__(self, et: ExtractedText):
+        self.et = et
+        self.step_size = 1024
+        self.index: dict[int, int] = self.create_index()
+
+    def create_index(self) -> dict[int, int]:
+        # We assume that the dictionary will preserve the order of insertion
+        # (Python 3.7+), so we can use it to create an index
+        # that maps the position of the text in the ExtractedText to the byte
+        # position in the serialized ExtractedText.
+        index: dict[int, int] = {}
+        encoded = self.et.SerializeToString()
+        # Initialize the index with the start position by searching for the first two characters.
+        # This is a trick to know where the protobuff header ends.
+        text_starts_at = encoded.index(self.et.text[:2].encode())
+        index[0] = text_starts_at
+
+        for i in range(self.step_size, len(self.et.text), self.step_size):
+            text_slice = self.et.text[0:i]
+            index[i] = len(text_slice.encode()) + text_starts_at
+
+        # Add the end position of the text to the index
+        last_pos = len(self.et.text) - 1
+        if last_pos not in index:
+            index[last_pos] = len(self.et.text.encode()) + text_starts_at
```

Comment:
I think this is equivalent to `index[last_pos] = encoded.len()`, assumming the protobuf does not have more fields, but not sure if that's a fair assumption.

If not, same comment as above, try to avoid reencoding the entire text once more.